File: lib/host/dailymotion.c
/* 
* Copyright (C) 2009,2010 Toni Gundogdu.
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "host.h"

_host_constants(dailymotion, "dailymotion.", "flv|hq|hd");

_host_re(re_id,    "(?i)video\\/(.*?)_");
_host_re(re_title, "(?i)title=\"(.*?)\"");
_host_re(re_path,  "(?i)%22(\\w\\w)URL%22%3A%22(.*?)%22");

QUVIcode
handle_dailymotion(const char *url, _quvi_video_t video) {
    static const char default_format[] = "sd";
    char *content, *best, *req_format;
    int ovector[30];
    QUVIcode rc;
    int offset;

    req_format =
        strcmp(video->quvi->format, "flv") == 0
        ? (char *)default_format
        : !is_format_supported(video->quvi->format, formats_dailymotion)
          ? (char *)default_format
          : video->quvi->format;

    /* host id */
    _host_id("dailymotion");

    /* common */
    rc = parse_page_common(url, video, &content, re_id, re_title);

    if (rc != QUVI_OK)
        return (rc);

    /*
     * can we dl this video? if it looks like a partner video,
     * walks like a partner video, then it's a partner video.
     * we don't care for those.
     */
    if (strstr(content, "SWFObject(\"http:") != 0) {
        _quvi_t quvi = video->quvi; /* seterr macro needs this. */
        seterr("looks like dailymotion partner video. refusing to continue");
        _free(content);
        return (QUVI_NOSUPPORT);
    }

    /* paths */
    offset = 0;
    best = NULL;
    do {
        char *path, *format_id;

        rc = regexp_capture(
            video->quvi,
            content+offset,
            re_path,
            ovector,
            sizeof(ovector)/sizeof(int),
            &format_id,
            &path,
            (void *) 0
        );

        if (rc == QUVI_OK) {

            path = strepl(path, "%5C", "");
            path = unescape(video->quvi, path);

            if (best != NULL)
                _free(best);

            asprintf(&best, "%s", path);

            if (strcmp(format_id, req_format) == 0)
                rc = add_video_link(&video->link, path);

            _free(path);
            _free(format_id);

            offset += ovector[1];
        }
    } while (rc == QUVI_OK);

    _free(content);

    if (best != NULL) {
        if (strcmp(req_format, "best") == 0)
            rc = add_video_link(&video->link, "%s", best);
        _free(best);
    }

    if (rc != QUVI_OK && !video->link)
        return (rc);

    return (QUVI_OK);
}

#ifdef _1_
static QUVIcode
parse(const _quvi_video_t video, char *paths) {
    static const char default_format[] = "spark";
    int best_width, best_flag;
    const char *format;
    char *best_link;
    char *s, *t;
    QUVIcode rc;

    format = 
        !strcmp(video->quvi->format, "flv")
        ? default_format /* flv -> spark which is dmotion's default fmt */
        : !is_format_supported(video->quvi->format, formats_dailymotion)
          ? default_format /* -> spark if requested fmt unsupported */
          : video->quvi->format;

    best_width = 0;
    best_flag  = !strcmp(format, "best");
    best_link   = 0;

    s = paths;
    t = strtok(s, "||");

    while (t) {
        char *path, *id;

        rc = regexp_capture(
            video->quvi,
            t,
            "(.*?)@@(.*?)$",
            0,
            0,
            &path,
            &id,
            (void *) 0
        );

        if (rc != QUVI_OK)
            return (rc);

        if (!best_flag) {
            if (!strcmp(id,format)) {
                rc = add_video_link(&video->link, "%s", path);
                _free(path);
                _free(id);
                return (rc);
            }
        }
        else { /* -f best */
            char *w, *h;

            rc = regexp_capture(
                video->quvi,
                path,
                "-(\\d+)x(\\d+)",
                0,
                0,
                &w,
                &h,
                (void *) 0
            );

            if (rc != QUVI_OK) {
                _free(path);
                _free(id);
                return (rc);
            }

            if (atoi(w) > best_width) {
                best_width = atoi(w);
                _free(best_link);
                asprintf(&best_link, "%s", path);
            }

            _free(w);
            _free(h);
        }

        _free(path);
        _free(id);

        t = strtok(0, "||");
    }

    rc = add_video_link(&video->link, "%s", best_link);

    _free(best_link);

    return (rc);
}

QUVIcode
handle_dailymotion(const char *url, _quvi_video_t video) {
    char *content, *paths;
    QUVIcode rc;

    /* host id */
    _host_id("dailymotion");

    /* common */
    rc = parse_page_common(url, video, &content, re_id, re_title);

    if (rc != QUVI_OK)
        return (rc);

    /* paths */
    rc = regexp_capture(
        video->quvi,
        content,
        re_paths,
        0,
        0,
        &paths,
        (void *) 0
    );

    _free(content);

    if (rc != QUVI_OK)
        return (rc);

    paths = unescape(video->quvi, paths); /* frees paths */

    rc = parse(video, paths);

    _free(paths);

    return (rc);
}
#endif

File: share/lua/website/dailymotion.lua

-- quvi
-- Copyright (C) 2010,2011  Toni Gundogdu <legatvs@gmail.com>
--
-- This file is part of quvi <http://quvi.sourceforge.net/>.
--
-- This library is free software; you can redistribute it and/or
-- modify it under the terms of the GNU Lesser General Public
-- License as published by the Free Software Foundation; either
-- version 2.1 of the License, or (at your option) any later version.
--
-- This library is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
-- Lesser General Public License for more details.
--
-- You should have received a copy of the GNU Lesser General Public
-- License along with this library; if not, write to the Free Software
-- Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
-- 02110-1301  USA
--

-- "http://dai.ly/cityofscars",
-- "http://www.dailymotion.com/video/xdpig1_city-of-scars_shortfilms",

local Dailymotion = {} -- Utility functions unique to this script.

-- Identify the script.
function ident(self)
    package.path = self.script_dir .. '/?.lua'
    local C      = require 'quvi/const'
    local r      = {}
    r.domain     = "dailymotion."
    r.formats    = "default|best"
    r.categories = C.proto_http
    local U      = require 'quvi/util'
    r.handles    = U.handles(self.page_url, {r.domain, "dai.ly"},
                    {"/video/","/%w+$","/family_filter"})
    return r
end

-- Query available formats.
function query_formats(self)
    local U       = require 'quvi/util'
    local page    = Dailymotion.fetch_page(self, U)
    local formats = Dailymotion.iter_formats(page, U)

    local t = {}
    for _,v in pairs(formats) do
        table.insert(t, Dailymotion.to_s(v))
    end

    table.sort(t)
    self.formats = table.concat(t, "|")

    return self
end

-- Parse media URL.
function parse(self)
    self.host_id = "dailymotion"

    local U    = require 'quvi/util'
    local page = Dailymotion.fetch_page(self, U)

    local _,_,s = page:find('title="(.-)"')
    self.title  = s or error("no match: media title")

    local _,_,s = page:find("video/(.-)_")
    self.id     = s or error("no match: media id")

    local _,_,s = page:find('"og:image" content="(.-)"')
    self.thumbnail_url = s or ''

    local formats = Dailymotion.iter_formats(page, U)
    self.url      = {U.choose_format(self, formats,
                                     Dailymotion.choose_best,
                                     Dailymotion.choose_default,
                                     Dailymotion.to_s).url
                        or error("no match: media url")}

    return self
end

--
-- Utility functions
--

function Dailymotion.fetch_page(self, U)
    self.page_url = Dailymotion.normalize(self.page_url)

    local _,_,s = self.page_url:find('/family_filter%?urlback=(.+)')
    if s then
        self.page_url = 'http://dailymotion.com' .. U.unescape(s)
    end

    local opts = {arbitrary_cookie = 'family_filter=off'}
    return quvi.fetch(self.page_url, opts)
end

function Dailymotion.normalize(page_url) -- "Normalize" embedded URLs
    if page_url:find("/swf/") then
        page_url = page_url:gsub("/video/", "/")
        page_url = page_url:gsub("/swf/", "/video/")
    end
    return page_url
end

function Dailymotion.iter_formats(page, U)
    local _,_,seq = page:find('"sequence",%s+"(.-)"')
    if not seq then
        local e = "no match: sequence"
        if page:find("_partnerplayer") then
            e = e .. ": looks like a partner video which we do not support"
        end
        error(e)
    end
    seq = U.unescape(seq)

    local _,_,msg = seq:find('"message":"(.-)[<"]')
    if msg then
        msg = msg:gsub('+',' ')
        error(msg:gsub('\\',''))
    end

    local _,_,vpp = seq:find('"videoPluginParameters":{(.-)}')
    if not vpp then
        -- See also <http://sourceforge.net/apps/trac/clive/ticket/4>
        error("no match: video plugin params")
    end

    local t = {}
    for url in vpp:gfind('%w+URL":"(.-)"') do
        url = url:gsub("\\/", "/")
        local _,_,c,w,h,cn = url:find('(%w+)%-(%d+)x(%d+).-%.(%w+)')
        if not c then
            error('no match: codec, width, height, container')
        end
--        print(c,w,h,cn)
        table.insert(t, {width=tonumber(w), height=tonumber(h),
                         container=cn,      codec=string.lower(c),
                         url=url})
    end

    return t
end

function Dailymotion.choose_default(formats) -- Lowest quality available
    local r = {width=0xffff, height=0xffff, url=nil}
    local U = require 'quvi/util'
    for _,v in pairs(formats) do
        if U.is_lower_quality(v,r) then
            r = v
        end
    end
--    for k,v in pairs(r) do print(k,v) end
    return r
end

function Dailymotion.choose_best(formats) -- Highest quality available
    local r = {width=0, height=0, url=nil}
    local U = require 'quvi/util'
    for _,v in pairs(formats) do
        if U.is_higher_quality(v,r) then
            r = v
        end
    end
--    for k,v in pairs(r) do print(k,v) end
    return r
end

function Dailymotion.to_s(t)
    return string.format("%s_%sp", t.container, t.height)
end

-- vim: set ts=4 sw=4 tw=72 expandtab:
